crypto/merkletrie/node.go

```diff
@@ -0,0 +1,383 @@
+
+		curChildNode.leaf = true
+		curChildNode.hash = n.hash[idiff+1:]
+		newChildNode.leaf = true
+		newChildNode.hash = d[idiff+1:]
+
+		pnode, nodeID = cache.allocateNewNode()
+		pnode.leaf = false
+		pnode.children = make([]storedNodeIdentifier, 256)
+		pnode.childrenNext = make([]byte, 256)
+
+		pnode.children[n.hash[idiff]] = curChildNodeID
+		pnode.children[d[idiff]] = newChildNodeID
+		if n.hash[idiff] < d[idiff] {
+			pnode.firstChild = n.hash[idiff]
+			pnode.childrenNext[pnode.firstChild] = d[idiff]
+			pnode.childrenNext[d[idiff]] = d[idiff]
+		} else {
+			pnode.firstChild = d[idiff]
+			pnode.childrenNext[pnode.firstChild] = n.hash[idiff]
+			pnode.childrenNext[n.hash[idiff]] = n.hash[idiff]
+		}
+		pnode.hash = append(path, d[:idiff]...)
+
+		for i := idiff - 1; i >= 0; i-- {
+			// create a parent node for pnode.
+			pnode2, nodeID2 := cache.allocateNewNode()
+			pnode2.leaf = false
+			pnode2.children = make([]storedNodeIdentifier, 256)
+			pnode2.childrenNext = make([]byte, 256)
+			pnode2.children[d[i]] = nodeID
+			pnode2.firstChild = d[i]
+			pnode2.childrenNext[d[i]] = d[i]
+			pnode2.hash = append(path, d[:i]...)
+
+			pnode = pnode2
+			nodeID = nodeID2
+		}
+		return nodeID, nil
+	}
+
+	if n.children[d[0]] == storedNodeIdentifierNull {
+		// no such child.
+		var childNode *node
+		var childNodeID storedNodeIdentifier
+		childNode, childNodeID = cache.allocateNewNode()
+		childNode.leaf = true
+		childNode.hash = d[1:]
+
+		pnode, nodeID = n.duplicate(cache)
+		pnode.children[d[0]] = childNodeID
+
+		if pnode.firstChild > d[0] {
+			pnode.childrenNext[d[0]] = pnode.firstChild
+			pnode.firstChild = d[0]
+		} else {
+			// iterate on all the entries.
+			i := pnode.firstChild
+			for {
+				if pnode.childrenNext[i] < d[0] {
+					if pnode.childrenNext[i] == i {
+						pnode.childrenNext[i] = d[0]
+						pnode.childrenNext[d[0]] = d[0]
+						break
+					}
+				} else {
+					pnode.childrenNext[d[0]] = pnode.childrenNext[i]
+					pnode.childrenNext[i] = d[0]
+					break
+				}
+				i = pnode.childrenNext[i]
+			}
+		}
+	} else {
+		// there is already a child there.
+		childNode, err := cache.getNode(n.children[d[0]])
+		if err != nil {
+			return storedNodeIdentifierNull, err
+		}
+		updatedChild, err := childNode.add(cache, d[1:], append(path, d[0]))
+		if err != nil {
+			return storedNodeIdentifierNull, err
+		}
+		pnode, nodeID = n.duplicate(cache)
+		cache.deleteNode(n.children[d[0]])
+		pnode.children[d[0]] = updatedChild
+	}
+	pnode.hash = path
+	return nodeID, nil
+}
+
+// calculateHash calculate the hash of the non-leaf nodes
+// when this function is called, the hashes of all the child node are expected
+// to have been calculated already. This is achived by doing the following:
+// 1. all node id allocations are done in incremental monolitic order, from the bottom up.
+// 2. hash calculations are being doing in node id incremental ordering
+func (n *node) calculateHash(cache *merkleTrieCache) error {
```

Comment:
Is the following correct?
Hash of a leaf node is the suffix of the value it represents that isn't already determined by its position in the trie. (i.e., path to leaf node + hash of leaf node = value the leaf node represents)
Hash of a middle node is crypto.Hash of the following:
* the length of the path to this node, as a byte
* the path to this node
* for each child node:
    * the length of the hash of that child + 1 if the child is not a leaf, or the length of the hash of that child + 128 if it is a leaf
    * the index of that child, a byte (i.e, the next byte on the path to the child)
    * the hash of the child

Is my understanding correct?

Reply:
almost.
two more things to note: 
1. In the for each child node, it's also according to the incremental order of the child node index.
2. the length of the element in the trie is always identical. It means that all leaf children would have the precise same length.

Reply:
I don't think this gives us the security properties we want. Say you give me a trie with RootHash `h` where the root node is not a leaf. I can construct a different trie that also has RootHash `h` as follows: the root node is a leaf, and the value it represents is (the hash you get from calling `calculateHash` on the root node of the trie you gave me). If I've understood correctly these tries will then both have the same RootHash.

Is it required that `calculateHash` leave the hash of leaf nodes unchanged? If it's allowed to change the hash of leaf nodes you could instead do something like the following:
* Hash of a leaf node is `concat(0, value it represents)`
* Hash of a non-leaf node node is `concat(1, (what calculateHash already does but without lenMod))`
This way node.hash can't be the same for leaves and non-leaves.

If you're not allowed to modify the hash of leaf nodes, you can have trie.RootHash distinguish between when the root is a leaf and when it isn't (e.g. return `crypto.Hash(concat(pnode.leaf ? 0 : 1, pnode.hash))`). I think also in `calculateHash` instead of leaves adding 128 vs 1 to the length field, there needs to be a separate byte for whether it's a leaf or not. Otherwise I can cause wraparound by giving you a tree with leaves of length `127+(length of a crypto.Digest)`. Element length isn't committed to in the RootHash, and the wraparound will make the parents of my leaves hash as if my nodes were non-leaves, so I'm pretty sure this can lead to a collision.

Reply:
updated.